db/versions/0010.yml

```diff
@@ -0,0 +1,379 @@
+          worker_pools.description,
+          worker_pools.config,
+          worker_pools.created,
+          worker_pools.last_modified,
+          worker_pools.owner,
+          worker_pools.email_on_error,
+          worker_pools.provider_data
+        from worker_pools
+        where worker_pools.worker_pool_id = worker_pool_id_in;
+      end
+  get_worker_pools:
+    description: |-
+      Get existing worker pools, ordered by `worker_pool_id`.  If the pagination arguments are both NULL, all rows are returned.
+      Otherwise, page_size rows are returned at offset page_offset.
+    mode: read
+    serviceName: worker_manager
+    args: page_size_in integer, page_offset_in integer
+    returns: table(worker_pool_id text, provider_id text, previous_provider_ids jsonb, description text, config jsonb, created timestamptz, last_modified timestamptz, owner text, email_on_error boolean, provider_data jsonb)
+    body: |-
+      begin
+        return query
+        select
+          worker_pools.worker_pool_id,
+          worker_pools.provider_id,
+          worker_pools.previous_provider_ids,
+          worker_pools.description,
+          worker_pools.config,
+          worker_pools.created,
+          worker_pools.last_modified,
+          worker_pools.owner,
+          worker_pools.email_on_error,
+          worker_pools.provider_data
+        from worker_pools
+        order by worker_pool_id
+        limit case
+            when (page_size_in is not null and page_size_in > 0) then page_size_in
+            else NULL
+          end
+        offset case
+            when (page_offset_in is not null and page_offset_in > 0) then page_offset_in
+            else 0
+          end;
+      end
+  update_worker_pool:
+    description: |-
+      Update API-accessible columns on an existig worker pool.  All fields are
+      overridden, but if the provider_id changes, then the existing provider_id
+      is added to previous_provider_ids.  The return value contains values
+      required for an API response and previous_provider_id (singular) containing
+      the provider_id found before the update.  If no such worker pool exists,
+      the return value is an empty set.
+    mode: write
+    serviceName: worker_manager
+    args: worker_pool_id_in text, provider_id_in text, description_in text, config_in jsonb, last_modified_in timestamptz, owner_in text, email_on_error_in boolean
+    returns: table(worker_pool_id text, provider_id text, description text, config jsonb, created timestamptz, last_modified timestamptz, owner text, email_on_error boolean, previous_provider_id text)
+    body: |-
+      declare
+        existing record;
+      begin
+        select
+          worker_pools.provider_id,
+          worker_pools.previous_provider_ids
+        from worker_pools
+        where worker_pools.worker_pool_id = worker_pool_id_in
+        -- lock this row for the duration of this transaction..
+        for update
+        into existing;
```

Comment:
Is this basically the same thing as using update on conflict?

Reply:
No, because in this case we're not doing an insert but an update -- but I could probably accomplish this with a single update query.  I'll give it a shot.

Reply:
Hm, https://www.postgresql.org/docs/9.4/dml-returning.html says
> In an UPDATE, the data available to RETURNING is the new content of the modified row.
so I think we wouldn't be able to return the `previous_provider_id` without selecting it first.  However, I think I can turn the update and final select into a single `update .. returning`.